core/domain/exp_domain.py

```diff
@@ -3519,3 +3519,226 @@ def to_metadata_dict(self):
             'title': self.title,
             'objective': self.objective,
         }
+
+
+class StateIdMapping(object):
+    """Domain object for state id mapping model."""
+
+    def __init__(
+            self, exploration_id, exploration_version, state_name_to_ids,
+            latest_state_id_used):
+        """Initialize state id mapping model object.
+
+        Args:
+            exploration_id: str. The exploration id whose state names are
+            mapped.
+            exploration_version: int. The version of the exploration.
+            state_name_to_ids: dict. A dictionary mapping a state name to
+                unique ID.
+            latest_state_id_used: int. Last used state id.
+        """
+        self.exploration_id = exploration_id
+        self.exploration_version = exploration_version
+        self.state_name_to_ids = state_name_to_ids
+        self.latest_state_id_used = latest_state_id_used
+
+    @staticmethod
+    def has_state_changed_significantly(old_state, new_state):
+        """Checks whether there is any significant change in old and new version
+        of the given state.
+
+        A siginificant change has happend when interaction id has been changed
+        in new version of the state.
+
+        Args:
+            old_state: dict. Old version of the state.
+            new_state: dict. New version of the state.
+
+        Returns:
+            bool. A boolean indicating whether change is significant or not.
+        """
+        if old_state.interaction.id != new_state.interaction.id:
+            # State has changed significantly is interaction id in new state is
+            # different from old state.
+            return True
+
+    @classmethod
+    def get_state_id_mapping(
+            cls, old_exploration, new_exploration, change_list):
+        """Get state name to id mapping for new exploration's states.
+
+        Args:
+            old_exploration: Exploration. Old version of the exploration.
+            new_exploration: Exploration. New version of the exploration.
+            change_list: list(dict). List of changes between old and new
+                exploration.
+
+        Returns:
+            StateIdMapping. Domain object for state id mapping.
+        """
+        new_state_names = []
+        state_id_to_names = {}
+
+        # Get state id mapping model for old exploration.
+        model = cls.get_state_id_mapping_from_model(
+            exp_models.StateIdMappingModel.get_state_id_mapping_model(
```

Comment:
Domain objects should not depend on models, actually. They should just be plain Python objects. In exp_services you might have something like:

    def create_and_save_state_id_mapping(exploration_id, exploration_version):
        ...
        # Note that get_state_id_mapping() should also be defined in exp_services, similar to what we do
        # for ExplorationModel.
        old_state_id_mapping = get_state_id_mapping(exploration_id, exploration_version - 1)  

        new_state_id_mapping = old_state_id_mapping.create_mapping_for_new_version(new_exploration, ...)
        save_state_id_mapping(exploration_id, exploration_version, new_state_id_mapping)

(this is untested, just an example, and you might need more args in the functions than I've written -- but the point is that all the get/set from datastore should happen in the exp_services file).

Reply:
Refactored code a bit. Also moved reverted mapping to a separate function.